Allow registering custom categories through filter

ernilambar opened the pull request:
Fixes https://github.com/WordPress/plugin-check/issues/494

* Introduce filter wp_plugin_check_categories
* Add new method `get_check_categories()` in `Check_Categories` class
* Add test for related changes

Example implementation: https://github.com/ernilambar/fresh-check/blob/main/fresh-check.php

ernilambar committed:
Add filter for check categories

ernilambar committed:
Add test for filter of check categories

swissspidy commented on includes/Checker/Check_Categories.php:
```diff
@@ -46,13 +40,35 @@ public static function get_categories() {
+		 *
+		 * @since 1.0.2
+		 *
+		 * @param array $default_categories An array of check categories.
```
Nit: let’s explain that it‘s an associative array

ernilambar replied:
Sure. Would it make more sense following array structure like `block_categories_all`? It would be more future proof and we can easily add attributes for each category in the future if needed.

swissspidy replied:
I don't think we need that. If we ever need more complex data structure we can still add it later.

What I mean here is something like 
```suggestion
		 * @param array<string, string> $default_categories Associative array of category slugs to labels.
```

swissspidy commented on includes/Checker/Check_Categories.php:
```diff
@@ -46,13 +40,35 @@ public static function get_categories() {
+	 *
+	 * @return array An array of check categories.
+	 */
+	public static function get_check_categories() {
```
Confusing to have get_check_categories and get_categories methods. Can we rename them?

swissspidy replied:
* `get_categories`
* `get_category_slugs`
* `get_category_labels`

ernilambar committed:
Correct filter docs

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

ernilambar committed:
Rename category methods

swissspidy commented on includes/Checker/Check_Categories.php:
```diff
@@ -46,13 +40,35 @@ public static function get_categories() {
 	 */
 	public static function get_category_labels() {
-		return array(
+		return self::get_categories();
```
```suggestion
		return array_values( self::get_categories() );
```

ernilambar committed:
Correct function for category labels

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

ernilambar commented:
Looks like we have used key/value pair, like:

```
$category_labels  = $check_categories->get_category_labels();

$categories = array();

foreach ( $category_labels as $slug => $label ) {
	$categories[] = array(
		'slug' => $slug,
		'name' => $label,
	);
}
```

So I think we need to adjust in other places also.

swissspidy commented:
Then these places should call `get_categories()`

swissspidy committed:
Merge branch 'trunk' into 494-customize-categories

ernilambar committed:
Update method name in list_check_categories() CLI command

ernilambar committed:
Update tests after method rename

mukeshpanchal27 commented on includes/Admin/Admin_Page.php:
```diff
@@ -194,7 +194,7 @@ public function render_page() {
 		$selected_plugin_basename = filter_input( INPUT_GET, 'plugin', FILTER_SANITIZE_FULL_SPECIAL_CHARS );
 
-		$category_labels = Check_Categories::get_category_labels();
+		$category_labels = Check_Categories::get_categories();
```
Update the variable to `$categories` and same `WP_PLUGIN_CHECK_PLUGIN_DIR_PATH . 'templates/admin-page.php'`

ernilambar committed:
Use categories instead of category_labels in admin page

mukeshpanchal27 committed:
Remove test_get_category_labels test

As `test_get_category_labels` check same function `get_categories()` let's remove it.

mukeshpanchal27 commented:
In https://github.com/WordPress/plugin-check/pull/509/commits/37aeb20e90194881d8b5cb497445866d8be5ada2, I have removed the `test_get_category_labels` tests as they test the same `get_categories()` data values, making them redundant.

mukeshpanchal27 committed:
Merge branch 'trunk' into 494-customize-categories

ernilambar commented:
May be we could remove that labels function. It is not used anywhere I guess.

mukeshpanchal27 commented:
Yes but that function introduce in 1.0, leave it for now.
